File: Java/06. Software Design Principles SOLID & More/Practice18.java
/**
 * Practice 18: Applying DIP (Dependency Inversion Principle)
 * Task: Build a payment system where PaymentService depends on abstraction,
 *       not concrete payment providers.
 *
 * How to compile and run:
 *   javac Practice18.java
 *   java Practice18
 */

import java.util.ArrayList;
import java.util.List;

public class Practice18 {

    interface PaymentMethod {
        String pay(double amount);
    }

    static class CreditCardPayment implements PaymentMethod {
        public String pay(double amount) {
            return String.format("Paid $%.2f using Credit Card", amount);
        }
    }

    static class BkashPayment implements PaymentMethod {
        public String pay(double amount) {
            return String.format("Paid $%.2f using bKash", amount);
        }
    }

    static class PaypalPayment implements PaymentMethod {
        public String pay(double amount) {
            return String.format("Paid $%.2f using PayPal", amount);
        }
    }

    static class BankTransferPayment implements PaymentMethod {
        public String pay(double amount) {
            return String.format("Paid $%.2f using Bank Transfer", amount);
        }
    }

    static class PaymentService {
        void processPayment(PaymentMethod paymentMethod, double amount) {
            if (amount <= 0) {
                throw new IllegalArgumentException("Amount must be greater than 0.");
            }
            System.out.println("[PaymentService] " + paymentMethod.pay(amount));
        }
    }

    public static void main(String[] args) {
        PaymentService service = new PaymentService();

        List<PaymentMethod> methods = new ArrayList<>();
        methods.add(new CreditCardPayment());
        methods.add(new BkashPayment());
        methods.add(new PaypalPayment());
        methods.add(new BankTransferPayment());

        for (PaymentMethod method : methods) {
            service.processPayment(method, 1500.0);
        }
    }
}

File: C#/06. Software Design Principles SOLID & More/Practice18.cs
/**
 * Practice 18: Applying DIP (Dependency Inversion Principle)
 * Task: Build a payment system where PaymentService depends on abstraction,
 *       not concrete payment providers.
 *
 * How to run (using dotnet-script):
 *   dotnet script Practice18.cs
 */

using System;
using System.Collections.Generic;

class Practice18
{
    interface IPaymentMethod
    {
        string Pay(decimal amount);
    }

    class CreditCardPayment : IPaymentMethod
    {
        public string Pay(decimal amount) => $"Paid ${amount:F2} using Credit Card";
    }

    class BkashPayment : IPaymentMethod
    {
        public string Pay(decimal amount) => $"Paid ${amount:F2} using bKash";
    }

    class PaypalPayment : IPaymentMethod
    {
        public string Pay(decimal amount) => $"Paid ${amount:F2} using PayPal";
    }

    class BankTransferPayment : IPaymentMethod
    {
        public string Pay(decimal amount) => $"Paid ${amount:F2} using Bank Transfer";
    }

    class PaymentService
    {
        public void ProcessPayment(IPaymentMethod paymentMethod, decimal amount)
        {
            if (amount <= 0)
                throw new ArgumentException("Amount must be greater than 0.");

            Console.WriteLine($"[PaymentService] {paymentMethod.Pay(amount)}");
        }
    }

    static void Main(string[] args)
    {
        var service = new PaymentService();
        var methods = new List<IPaymentMethod>
        {
            new CreditCardPayment(),
            new BkashPayment(),
            new PaypalPayment(),
            new BankTransferPayment()
        };

        foreach (var method in methods)
            service.ProcessPayment(method, 1500m);
    }
}
